### PR TITLE
HTTP: Add TLS version configurability for grafana server

### PR DESCRIPTION
**Pull request by venkatbvc**
Backend: Add TLS version configurability for grafana server. Also backward compatibility is supported. A new parameter min_tls_version are added to the [server] section.

**What is this feature?**

Currently TLS version and ciphers are hardcoded for http_server. This PR adds configurability from the user point of view.

**Why do we need this feature?**

User doesn't flexibility to add ciphers and fix version of TLS that needs to be used.

**Who is this feature for?**

For all

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by venkatbvc** [comment]
@torkelo  How do i add the following labels:
nobackport
milestone can be next minor version
Changelog - i have updated document. required document, but am not sure whether it needs to be part of changelog.md

**Comment by venkatbvc** [comment]
no-backport

**Comment by venkatbvc** [comment]
Hi All,
Can anyone check and let me know what needs to be done for adding the required label for checks to pass

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:194]:
```suggestion
The TLS Handshake requires a minimum TLS version. The available options are TLS1.2 and TLS1.3.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:195]:
```suggestion
If you do not specify a version, the system uses TLS1.2.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:200]:
```suggestion
For a list of supported ciphers and TLS versions, refer to [Source file src/crypto/tls/cipher_suites.go](https://go.dev/src/crypto/tls/cipher_suites.go).
```

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:199]:
```suggestion
This section refers to the values that are accepted for the TLS cipher. By default, this value is left empty. Example value: "TLS_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256"
```

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:203]:
```suggestion
If the protocol is set to HTTPS and there are no ciphers configured, the system will apply the following settings to support backward compatibility for TLS 1.2.: `TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,TLS_RSA_WITH_AES_128_GCM_SHA256,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_128_CBC_SHA,TLS_RSA_WITH_AES_256_CBC_SHA`
```

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:205]:
```suggestion
The following settings are applied for TLS1.2 when h2 is the configured protocol and no ciphers are configured:
```

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:206]:
```suggestion
`TLS_CHACHA20_POLY1305_SHA256,TLS_AES_128_GCM_SHA256,TLS_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305`
```

**Comment by chri2547** [comment]
@venkatbvc , thanks for the contribution. I added some docs suggestions; please review them before accepting. I want to ensure I didn't change the technical meaning. I've added a milestone and the no-changelog label. As for the backport label, I'll defer to engineering for that.

**Review by chri2547 — COMMENTED** [review]
added docs suggestions. Please ping me for a re-review. Thanks for the contribution!

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (venkatbvc) [reply docs/sources/setup-grafana/configure-grafana/_index.md]: instead of settings, we can put ciphers

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (venkatbvc) [reply docs/sources/setup-grafana/configure-grafana/_index.md]: This comment is fine. i will have made changes

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (venkatbvc) [reply docs/sources/setup-grafana/configure-grafana/_index.md]: I have made changes.

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (venkatbvc) [reply docs/sources/setup-grafana/configure-grafana/_index.md]: i have made changes

**Comment by venkatbvc** [comment]
@chri2547 Thanks. Your comments look fine. I have addressed them and published the changes.

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:202]:
```suggestion
```

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:203]:
```suggestion
If the protocol is set to HTTPS and there are no ciphers configured, the system will apply the following ciphers to support backward compatibility for TLS 1.2: `TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,TLS_RSA_WITH_AES_128_GCM_SHA256,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_128_CBC_SHA,TLS_RSA_WITH_AES_256_CBC_SHA`
```

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:202]:
```suggestion
#### Backward compatibility
```

**Review by chri2547 — COMMENTED** [review]
Two very small suggestions. Approving these docs changes.

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (venkatbvc) [reply docs/sources/setup-grafana/configure-grafana/_index.md:203]: i have made it small case, as the protocol that is configured in .ini file is small case. Let me know if you want to change?

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (venkatbvc) [reply docs/sources/setup-grafana/configure-grafana/_index.md:202]: will make changes

**Comment by venkatbvc** [comment]
@chri2547  Thanks for reviewing. Addressed the minor comments and published the changes. will wait for engineering team to review and share comments

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md]: ah, then we should use tickmarks, like `https`. I'll make that change and commit my suggestion (for expediency).

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:204]:
```suggestion
If the protocol is set to `https` and there are no ciphers configured, the system will apply the following ciphers to support backward compatibility for TLS 1.2: `TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,TLS_RSA_WITH_AES_128_GCM_SHA256,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_128_CBC_SHA,TLS_RSA_WITH_AES_256_CBC_SHA`
```

Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:204]:
```suggestion
If the protocol is set to`https` and there are no ciphers configured, the system will apply the following ciphers to support backward compatibility for TLS 1.2: `TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,TLS_RSA_WITH_AES_128_GCM_SHA256,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_128_CBC_SHA,TLS_RSA_WITH_AES_256_CBC_SHA`
```

File: docs/sources/setup-grafana/configure-grafana/_index.md
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:206]:
```suggestion
If the protocol is set to`h2` and there are no ciphers configured, the system will apply the following ciphers to support backward compatibility for TLS 1.2:
```

**Review by chri2547 — COMMENTED** [review]
Docs approved.

**Comment by venkatbvc** [comment]
@torkelo @suntala @zserge @danielkenlee @sakjur  Can you please review and provide your comments

**Comment by sakjur** [comment]
This is fairly low prio right now, it missed our Grafana 10.0 window and it'd at best make it into 10.1.

I don't know if I think we should do this at all, I'm all for configuring the minimum TLS version (after all, if all your users are using modern browsers, it makes sense to only permit TLS 1.3), but for exact ciphersuites, I'm skeptical. Partly because we're unable to configure the ciphersuites for TLS 1.3 because that's not supported by Go upstream (https://github.com/golang/go/issues/29349) and partly because detailed configuration of TLS is sort of what reverse proxies such as nginx and Caddy excels at.

Quick review:
If we're to merge this I don't think we should be linking into the Go codebase for documenting the available options (that list is unlikely to change very much, so I think we could copy it into our docs), and you should use `github.com/grafana/grafana/pkg/util.SplitString` for splitting the list items.

**Comment by TomsioncatiGraf** [comment]
Configurability of cipher suites
👍 Increased user optionality, normally good
👎 Go is very good at picking cipher suites, so generally bad for security overall

If this isn't a direct customer need, I'd suggest not making cipher suites configurable and trusting Go upstream

**Comment by venkatbvc** [comment]
@sakjur @TomsioncatiGraf   Thanks for your comments. I have started with only make tls version configurable. But when i checked at the ldap.toml these were made configurable. So i thought it may be a good idea to configure.
So if you think tls_ciphers better not to configure and leave what is configured in the code by default, i can change PR only to set min tls version

**Comment by sakjur** [comment]
@venkatbvc If we can start there, that’d be something I’d definitely approve 🙂

**Comment by venkatbvc** [comment]
@sakjur  Thanks. I will publish changes only for tls version. tls_ciphers i can do later if required

**Comment by venkatbvc** [comment]
@sakjur pushed code only for configuring tls version. Please check

**Comment by sakjur** [comment]
@venkatbvc Thank you! I'll put this on my todo to review 🙂

**Comment by rpasche** [comment]
2 questions, as I stumbled on this PR while trying to configure ciphers to be used.

1. I'm questioning, if this isn't an error? What, if I wanted to configure the minTLSVersion == 1.3? (https://github.com/grafana/grafana/pull/67482/files#diff-059f97ccda689ebea0e2f9981cd736d77b1fa01a719d95abae333ad40fb0b9c4R744)
2. Although we could then set a `minTLSVersion` to be configurable, the list of ciphers (attached to each version) still looks unchangeable to me. Or am I missing something?

**Comment by venkatbvc** [comment]
@rpasche  Initially, i have made tls version and ciphers configurable. Later, it was found that tlsCiphers setting have issues(Check this comment: https://github.com/grafana/grafana/pull/67482#issuecomment-1534405423) . So removed setting the TLS ciphers. only TLS version is configurable now.
If TLS version is set to 1.3, then crpto/tls will choose the default ciphers.

**Comment by rpasche** [comment]
> @rpasche Initially, i have made tls version and ciphers configurable. Later, it was found that tlsCiphers setting have issues(Check this comment: [#67482 (comment)](https://github.com/grafana/grafana/pull/67482#issuecomment-1534405423)) . So removed setting the TLS ciphers. only TLS version is configurable now. If TLS version is set to 1.3, then crpto/tls will choose the default ciphers. Following is the excerpt from crypto/tls:
> 
> // CipherSuites is a list of enabled TLS 1.0–1.2 cipher suites. The order of // the list is ignored. Note that TLS 1.3 ciphersuites are not configurable. // // If CipherSuites is nil, a safe default list is used. The default cipher // suites might change over time. CipherSuites [][uint16]

@venkatbvc thank you for this explanation.

**Review by sakjur — APPROVED** [review]
Tested this locally and it works great, thanks for your contribution 🎉

**Pull request merged**
